### PR TITLE
feat(ui-map,api): add districts panel and fix layer issues

### PR DESCRIPTION
**Pull request by hdinia**

### CODE REVIEW

File: webapp/src/components/App/Singlestudy/explore/Modelization/Map/MapConfig/Districts/CreateDistrictDialog.tsx [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    output: true,
+  };
+
+  ////////////////////////////////////////////////////////////////
+  // Event Handlers
+  ////////////////////////////////////////////////////////////////
+
+  const handleSubmit = (data: SubmitHandlerPlus<typeof defaultValues>) => {
+    const { name, output } = data.values;
+    dispatch(
```
Comment (skamril) [inline webapp/src/components/App/Singlestudy/explore/Modelization/Map/MapConfig/Districts/CreateDistrictDialog.tsx]: Use a snackbar for error

File: webapp/src/components/App/Singlestudy/explore/Modelization/Map/MapConfig/Districts/CreateDistrictDialog.tsx [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    onClose();
+  };
+
+  ////////////////////////////////////////////////////////////////
+  // JSX
+  ////////////////////////////////////////////////////////////////
+
+  return (
+    <FormDialog
+      title="Add new district"
```
Comment (skamril) [inline webapp/src/components/App/Singlestudy/explore/Modelization/Map/MapConfig/Districts/CreateDistrictDialog.tsx:61]: i18n

File: webapp/src/components/App/Singlestudy/explore/Modelization/Map/MapConfig/Districts/EditDistrictDialog.tsx [hunk 3]
```diff
@@ -0,0 +1,141 @@
+import { useTranslation } from "react-i18next";
```
Comment (skamril) [inline webapp/src/components/App/Singlestudy/explore/Modelization/Map/MapConfig/Districts/EditDistrictDialog.tsx]: UpdateDistrictDialog.jsx instead. For convenience we keep CRUD terms for all files.

File: webapp/src/components/App/Singlestudy/explore/Modelization/Map/MapConfig/Districts/CreateDistrictDialog.tsx [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+  onClose: () => void;
+}
+
+function CreateDistrictDialog(props: Props) {
+  const { open, onClose } = props;
+  const { study } = useOutletContext<{ study: StudyMetadata }>();
+  const [t] = useTranslation();
+  const dispatch = useAppDispatch();
+
+  const defaultValues = {
```
Comment (skamril) [inline webapp/src/components/App/Singlestudy/explore/Modelization/Map/MapConfig/Districts/CreateDistrictDialog.tsx]: Move outside the component

File: webapp/src/components/App/Singlestudy/explore/Modelization/Map/MapConfig/Districts/EditDistrictDialog.tsx [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+
+function EditDistrictDialog(props: Props) {
+  const { open, onClose } = props;
+  const { study } = useOutletContext<{ study: StudyMetadata }>();
+  const [t] = useTranslation();
+  const dispatch = useAppDispatch();
+  const districts = useAppSelector(getStudyMapDistricts);
+  const [openConfirmationModal, setOpenConfirmationModal] = useState(false);
+
+  const defaultValues = {
```
Comment (skamril) [inline webapp/src/components/App/Singlestudy/explore/Modelization/Map/MapConfig/Districts/EditDistrictDialog.tsx]: Move outside the component

File: webapp/src/components/App/Singlestudy/explore/Modelization/Map/MapConfig/Districts/EditDistrictDialog.tsx [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+    name: "",
+    output: true,
+    districtId: "",
+  };
+
+  const districtsOptions = Object.values(districts).map(({ name, id }) => ({
```
Comment (skamril) [inline webapp/src/components/App/Singlestudy/explore/Modelization/Map/MapConfig/Districts/EditDistrictDialog.tsx]: useMemo (you don't know how many districts there is)

File: webapp/src/redux/ducks/studyMaps.ts [hunk 7]
```diff
@@ -56,6 +56,14 @@ export interface StudyMapLayer {
   areas: StudyMapNode["id"][];
 }
 
+export interface StudyMapDistrict {
+  id: string;
+  name: string;
+  output: boolean;
+  comments: string;
+  areas: StudyMapNode["id"][];
```
Comment (skamril) [inline webapp/src/redux/ducks/studyMaps.ts]: `Array<StudyMapNode["id"]>` is more readable here

File: webapp/src/redux/ducks/studyMaps.ts [hunk 8]
```diff
@@ -56,6 +56,14 @@ export interface StudyMapLayer {
   areas: StudyMapNode["id"][];
```
Comment (skamril) [inline webapp/src/redux/ducks/studyMaps.ts]: `Array<StudyMapNode["id"]>` is more readable here

File: webapp/src/redux/ducks/studyMaps.ts [hunk 9]
```diff
@@ -345,6 +368,101 @@ export const deleteStudyMapLayer = createAsyncThunk<
...
+            name,
+            output,
+            comments,
+            areas,
+          };
+          return acc;
+        },
+        {} as StudyMapsState["districts"]
+      );
+      initStudyMapDistricts(dispatch, studyMapDistricts);
```
Comment (skamril) [inline webapp/src/redux/ducks/studyMaps.ts]: `return studyMapDistricts` (cf. https://github.com/AntaresSimulatorTeam/AntaREST/pull/1223#discussion_r1071482372)

File: webapp/src/redux/ducks/studyMaps.ts [hunk 10]
```diff
@@ -56,6 +56,14 @@ export interface StudyMapLayer {
   areas: StudyMapNode["id"][];
 }
 
+export interface StudyMapDistrict {
+  id: string;
+  name: string;
+  output: boolean;
+  comments: string;
```
Comment (skamril) [inline webapp/src/redux/ducks/studyMaps.ts:72]: `comment` in singular maybe?

File: webapp/src/redux/ducks/studyMaps.ts [hunk 11]
```diff
@@ -345,6 +368,101 @@ export const deleteStudyMapLayer = createAsyncThunk<
...
+        {} as StudyMapsState["districts"]
+      );
+      initStudyMapDistricts(dispatch, studyMapDistricts);
+    } catch (err) {
+      return rejectWithValue(err);
+    }
+  }
+);
+
+export const createStudyMapDistrict = createAsyncThunk<
```
Comment (skamril) [inline webapp/src/redux/ducks/studyMaps.ts:483]: `createAsyncThunk<StudyMapDistrict, ...`

File: webapp/src/redux/ducks/studyMaps.ts [hunk 12]
```diff
@@ -345,6 +368,101 @@ export const deleteStudyMapLayer = createAsyncThunk<
...
+  } catch (error) {
+    return rejectWithValue(error);
+  }
+});
+
+export const updateStudyMapDistrict = createAsyncThunk<
+  {
+    districtId: StudyMapDistrict["id"];
+    output: StudyMapDistrict["output"];
+    areas?: StudyMapNode[];
```
Comment (skamril) [inline webapp/src/redux/ducks/studyMaps.ts]: `nodes?:` is more appropriate here

File: webapp/src/redux/ducks/studyMaps.ts [hunk 13]
```diff
@@ -345,6 +368,101 @@ export const deleteStudyMapLayer = createAsyncThunk<
...
+  },
+  {
+    studyId: StudyMetadata["id"];
+    districtId: StudyMapDistrict["id"];
+    output: StudyMapDistrict["output"];
+    areas?: StudyMapNode[];
```
Comment (skamril) [inline webapp/src/redux/ducks/studyMaps.ts]: Same

File: webapp/src/redux/ducks/studyMaps.ts [hunk 14]
```diff
@@ -345,6 +368,101 @@ export const deleteStudyMapLayer = createAsyncThunk<
...
+    await studyApi.updateStudyDistrict(studyId, districtId, output, areas);
+    return { districtId, output, areas };
+  } catch (error) {
+    return rejectWithValue(error);
+  }
+});
+
+export const deleteStudyMapDistrict = createAsyncThunk<
+  {
+    districtId: StudyMapDistrict["id"];
```
Comment (skamril) [inline webapp/src/redux/ducks/studyMaps.ts:539]: Return the id directly: `createAsyncThunk<StudyMapDistrict["id"], ...`

File: webapp/src/redux/ducks/studyMaps.ts [hunk 15]
```diff
@@ -91,6 +101,10 @@ export const setLayers = createAction<
   NonNullable<Record<StudyMapLayer["id"], StudyMapLayer>>
 >(n("SET_LAYERS"));
 
+export const setDistricts = createAction<
```
Comment (skamril) [inline webapp/src/redux/ducks/studyMaps.ts]: To remove, `fetchStudyMapDistricts` is enough

File: webapp/src/redux/ducks/studyMaps.ts [hunk 16]
```diff
@@ -440,5 +558,34 @@ export default createReducer(initialState, (builder) => {
     .addCase(deleteStudyMapLayer.fulfilled, (draftState, action) => {
       const { layerId } = action.payload;
       delete draftState.layers[layerId];
+    })
+    .addCase(setDistricts, (draftState, action) => {
```
Comment (skamril) [inline webapp/src/redux/ducks/studyMaps.ts]: `.addCase(fetchStudyMapDistricts.fulfilled, ...` (cf. https://github.com/AntaresSimulatorTeam/AntaREST/pull/1223#discussion_r1071482372)

File: webapp/src/redux/ducks/studyMaps.ts [hunk 17]
```diff
@@ -440,5 +558,34 @@ export default createReducer(initialState, (builder) => {
...
+      const { districtId } = action.payload;
+      delete draftState.districts[districtId];
+    })
+    .addCase(updateStudyMapDistrict.fulfilled, (draftState, action) => {
+      const { districtId, output, areas } = action.payload;
+      if (areas) {
+        draftState.districts[districtId] = {
+          ...draftState.districts[districtId],
+          output,
+          // eslint-disable-next-line @typescript-eslint/ban-ts-comment
```
Comment (skamril) [inline webapp/src/redux/ducks/studyMaps.ts]: Remove comments and `[...areas]` is not necessary

File: webapp/src/redux/selectors.ts [hunk 18]
```diff
@@ -467,6 +467,12 @@ export const getStudyMapLinks = createSelector(
   }
 );
 
+export const getStudyMapDistricts = (
+  state: AppState
+): StudyMapsState["districts"] => {
+  return getStudyMapsState(state).districts;
```
Comment (skamril) [inline webapp/src/redux/selectors.ts:473]: Rename `getStudyMapDistrictsById` or create a selector with Object.values

**Review by skamril — CHANGES_REQUESTED** [review]

File: webapp/src/components/App/Singlestudy/explore/Modelization/Map/MapConfig/Districts/CreateDistrictDialog.tsx (hunk 2 above)
Comment (hdinia) [reply webapp/src/components/App/Singlestudy/explore/Modelization/Map/MapConfig/Districts/CreateDistrictDialog.tsx:61]: not needed here

File: webapp/src/components/App/Singlestudy/explore/Modelization/Map/MapConfig/Districts/EditDistrictDialog.tsx (hunk 6 above)
Comment (hdinia) [reply webapp/src/components/App/Singlestudy/explore/Modelization/Map/MapConfig/Districts/EditDistrictDialog.tsx]: Object.values() and Array.map() are already very efficient and the districts object is not likely to change frequently, so the performance benefits of using the useMemo hook would be minimal. Not really necessary here I think

File: webapp/src/redux/ducks/studyMaps.ts (hunk 10 above)
Comment (hdinia) [reply webapp/src/redux/ducks/studyMaps.ts:72]: I agree `comment` is better but this is what is expected by the api: `comments`

File: webapp/src/redux/ducks/studyMaps.ts (hunk 12 above)
Comment (hdinia) [reply webapp/src/redux/ducks/studyMaps.ts]: `areas` is expected, the type is not very descriptive though. `StudyMapDistrict["areas"]` would be more correct here, maybe?

File: webapp/src/redux/ducks/studyMaps.ts (hunk 17 above)
Comment (hdinia) [reply webapp/src/redux/ducks/studyMaps.ts]: Thx mybad

**Pull request merged**